### DISCUSSION TITLE
Implementing the Inspector API

### DISCUSSION THREAD

**Discussion by shrirambalaji** [jstime/jstime#64] [Ideas]
Based on a previous discussion in https://github.com/jstime/jstime/pull/62 @devsnek  and @MylesBorins mentioned adding an Inspector API to JSTime seemed like a good idea. It would be great to discuss upon how a JSTime Inspector would look like. After going through some [articles](https://hyperandroid.com/2020/02/12/v8-inspector-from-an-embedder-standpoint/) I have a very basic understanding of V8's Inspector.

## Questions:

- What are the responsibilities of the JSTime Inspector? Should it exist simply as an `impl` that just wraps around V8 Inspector, taking in the isolate_state? Something like:

  ```
  // core/src/inspector.rs

  // Heavily based on Rusty V8's unit tests
  // https://github.com/denoland/rusty_v8/blob/e5f56e218c68e6910fc610b8e99d7b08cf6619b4/tests/test_api.rs#L3134

  impl InspectorClient {
      pub fn new() -> Self {
          Self {
              base: V8InspectorClientBase::new::<Self>(),
          }
      }
  }

  // I don't understand what V8InspectorClientBase does tbh
  // but it seems to be required for creating a client.

  impl V8InspectorClientImpl for InspectorClient {
      fn base(&self) -> &V8InspectorClientBase {
          &self.base
      }

      fn base_mut(&mut self) -> &mut V8InspectorClientBase {
          &mut self.base
      }
  }


  impl Inspector {
      pub fn new(isolate: &mut v8::OwnedIsolate) -> Self {
          let mut client = InspectorClient::new();
          let inspector = Self {
              inspector: V8Inspector::create(isolate, &mut client),
          };
          inspector
      }

  ```

  > I might need to actually use `IsolateState.get` here, since isolate cannot (and should not) be mutably borrowed more than once.

- Should we mimic Deno as-is? 
  [Deno's Inspector API](https://github.com/denoland/deno/blob/master/cli/inspector.rs) is written as part of CLI mod, I'm not sure if we want it to be part of JSTime Core. Also Deno's implementation seems to implement the V8 Inspector Client, and create a websocket communication channel for Chrome DevTools ( which JSTime might also need to do at some point for devtools compat?).

  For the purposes of #62 however adding compat to communicate over the Inspector Protocol doesn't seem necessary.

- How would the inspector be used directly for #62? Would an `inspector::inspect` method run an `eval` on an expression and return a `Result<String, String>`? 

- Is there a way to get contextual information about a statement / expression directly using the Inspector. For eg. in `let a = 20;`  would it return information like whether `a` is a `VariableDeclaration`? 

> Still a rust / v8 noob, apologise for any technical inaccuracies above :smile:

**Comment by devsnek** [comment]
> How would the inspector be used directly for #62? Would an inspector::inspect method run an eval on an expression and return a Result<String, String>?

Grep for `session.post` in https://github.com/nodejs/repl/blob/master/src/index.js. Those are raw V8 inspector commands.

> Is there a way to get contextual information about a statement / expression directly using the Inspector. For eg. in let a = 20; would it return information like whether a is a VariableDeclaration?

No

**Reply by shrirambalaji** [reply]
@devsnek got it, so the repl needs to create a WS channel, and that would encapsulate an `InspectorSession`. Similar to node-repl we need to handle the messages in the session (maybe like this https://github.com/denoland/deno/blob/master/cli/inspector.rs) directly from V8.

Also do you want the Inspector to be part of core, or the CLI module?

**Reply by devsnek** [reply]
i think core.
